modules/apps/mail/mail-settings-test/src/testIntegration/java/com/liferay/mail/settings/internal/upgrade/v1_0_0/test/MailSettingCompanyConfigurationUpgradeProcessTest.java

```diff
@@ -0,0 +1,223 @@
+
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+
+/**
+ * @author Jiefeng Wu
+ */
+@RunWith(Arquillian.class)
+public class MailSettingCompanyConfigurationUpgradeProcessTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new LiferayIntegrationTestRule();
+
+	@BeforeClass
+	public static void setUpClass() throws Exception {
+		_storedMailSettingCompanyConfiguration =
+			_configurationAdmin.getConfiguration(
+				"com.liferay.mail.settings.configuration." +
+					"MailSettingCompanyConfiguration",
+				StringPool.QUESTION);
+
+		_storedProperties =
+			_storedMailSettingCompanyConfiguration.getProperties();
+
+		_configurationProvider.deleteSystemConfiguration(
+			MailSettingCompanyConfiguration.class);
+	}
+
+	@AfterClass
+	public static void tearDownClass() throws Exception {
+		_configurationProvider.saveSystemConfiguration(
+			MailSettingCompanyConfiguration.class, _storedProperties);
+	}
+
+	@Before
+	public void setUp() throws Exception {
+		_upgradeProcess = UpgradeTestUtil.getUpgradeStep(
+			_upgradeStepRegistrator,
+			"com.liferay.mail.settings.internal.upgrade.v1_0_0." +
+				"MailSettingCompanyConfigurationUpgradeProcess");
+
+		_testCompanyIdPortletPreferences = _prefsProps.getPreferences(
+			TestPropsValues.getCompanyId());
+
+		_testSystemIdPortletPreferences = _prefsProps.getPreferences();
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		_configurationProvider.deleteCompanyConfiguration(
+			MailSettingCompanyConfiguration.class,
+			TestPropsValues.getCompanyId());
+		_configurationProvider.deleteSystemConfiguration(
+			MailSettingCompanyConfiguration.class);
+	}
+
+	@Test
+	public void testUpgradeMailSettingCompanyConfigurationWithCompanyId()
+		throws Exception {
+
+		String stringValue = RandomTestUtil.randomString();
+		String intValue = String.valueOf(RandomTestUtil.randomInt());
+
+		_populatePreferences(
+			_testCompanyIdPortletPreferences, stringValue, intValue);
+
+		_upgradeProcess.upgrade();
+
+		Thread.sleep(2000);
+
+		MailSettingCompanyConfiguration mailSettingCompanyConfiguration =
+			_configurationProvider.getCompanyConfiguration(
+				MailSettingCompanyConfiguration.class,
+				TestPropsValues.getCompanyId());
+
+		_assertConfiguration(
+			stringValue, intValue, mailSettingCompanyConfiguration);
+	}
+
+	@Test
+	public void testUpgradeMailSettingCompanyConfigurationWithSystemId()
+		throws Exception {
+
+		String stringValue = RandomTestUtil.randomString();
+		String intValue = String.valueOf(RandomTestUtil.randomInt());
+
+		_populatePreferences(
+			_testSystemIdPortletPreferences, stringValue, intValue);
+
+		_upgradeProcess.upgrade();
+
+		Thread.sleep(2000);
```

Comment:
@jeffwu0724 What is this sleep for?

Reply:
Hi @tinatian ,

This is because during the ci test, after we finished the upgrade process, the `mailSettingCompanyConfiguration` I got from the `_configurationProvider` sometimes did not have the updated osgi config, and it will fail the test.  Therefore, I add a sleep here to wait for the `mailSettingCompanyConfiguration` to refresh to get the latest osgi config applied from the upgrade process.

I can see this issue only on ci, and locally, the test is always passing without this sleep.